devicetypes/smartthings/zigbee-power-meter.src/zigbee-power-meter.groovy

```diff
@@ -72,29 +75,31 @@ def parse(String description) {
         def descMap = zigbee.parseDescriptionAsMap(description)
         log.debug "Desc Map: $descMap"
                 
-        List attrData = [[clusterInt: descMap.clusterInt ,attrInt: descMap.attrInt, value: descMap.value]]
+        List attrData = [[clusterInt: descMap.clusterInt ,attrInt: descMap.attrInt, value: descMap.value, isValidForDataType: descMap.isValidForDataType]]
```

Comment:
Can You explain why isValidForDataType: descMap.isValidForDataType is added here ?

Reply:
Konrad, some devices report invalid data (e.g. during start up or power on). We need to check if the data being sent by the device is valid before processing it otherwise in the next line of code `zigbee.convertHexToInt(it.value)` it will take that invalid data and report an invalid value to the platform. The platform ZigBee libraries do the validation of data sent and then sets the `isValidForDataType` field for each reported attribute in `parseDescriptionAsMap`. This field should to be checked to avoid invalid data from being processed in the DTH, hence it's saved in the attrData by the DTH. Here's an example of data reported by the device which is invalid and if processed it results in an arbitrary report to the platform:

> 2cbda69a-bc3f-49f5-bbeb-bbff99ec8235 2:33:46 PM: debug Parse returned [name:energy, value:281474976.710655, unit:kWh, linkText:frient Energy Monitor, descriptionText:frient Energy Monitor energy is 281474976.710655 kWh, isStateChange:true, displayed:true]
2cbda69a-bc3f-49f5-bbeb-bbff99ec8235 2:33:46 PM: debug energy
2cbda69a-bc3f-49f5-bbeb-bbff99ec8235 2:33:46 PM: debug Desc Map: [raw:BF5102070226000025FFFFFFFFFFFF00042A08000000021800, dni:BF51, endpoint:02, cluster:0702, size:38, attrId:0000, result:success, encoding:25, value:ffffffffffff, additionalAttrs:[[attrId:0400, attrInt:1024, encoding:2A, value:000008, isValidForDataType:true, consumedBytes:12], [attrId:0200, attrInt:512, encoding:18, value:00, isValidForDataType:true, consumedBytes:8]], isValidForDataType:false, clusterInt:1794, attrInt:0]
2cbda69a-bc3f-49f5-bbeb-bbff99ec8235 2:33:46 PM: warn Received invalid data for type. AttrId: 0x0000, Type: 25, Value: ffffffffffff
2cbda69a-bc3f-49f5-bbeb-bbff99ec8235 2:33:46 PM: debug description is read attr - raw: BF5102070226000025FFFFFFFFFFFF00042A08000000021800, dni: BF51, endpoint: 02, cluster: 0702, size: 38, attrId: 0000, result: success, encoding: 25, value: 001802000000082a0400ffffffffffff

Reply:
All done

Reply:
@greens What do You think about it ?  Can we expect descMap.isValidForDataType in the map returned from parseDescriptionAsMap() in every case?

Reply:
I'll say it probably doesn't hurt.